service/matching/matcher.go

```diff
@@ -0,0 +1,210 @@
+	// are interested in queryTasks but not others. Example is when domain is
+	// not active in a cluster
+	queryTaskC chan *internalTask
+	// ratelimiter that limits the rate at which tasks can be dispatched to consumers
+	limiter *rateLimiter
+	// domain metric scope
+	scope func() metrics.Scope
+}
+
+const (
+	_defaultTaskDispatchRPS    = 100000.0
+	_defaultTaskDispatchRPSTTL = 60 * time.Second
+	// maxRateLimitWaitTime is the max amount of time that we are willing to wait for a
+	// ratelimit token to be available in future
+	maxRateLimitWaitTime = int64(200 * time.Millisecond)
+)
+
+var errTasklistThrottled = errors.New("cannot add to tasklist, limit exceeded")
+
+// newTaskMatcher returns an task matcher instance. The returned instance can be
+// used by task producers and consumers to find a match. Both sync matches and non-sync
+// matches should use this implementation
+func newTaskMatcher(config *taskListConfig, scopeFunc func() metrics.Scope) *TaskMatcher {
+	dPtr := _defaultTaskDispatchRPS
+	limiter := newRateLimiter(&dPtr, _defaultTaskDispatchRPSTTL, config.MinTaskThrottlingBurstSize())
+	return &TaskMatcher{
+		limiter:    limiter,
+		scope:      scopeFunc,
+		taskC:      make(chan *internalTask),
+		queryTaskC: make(chan *internalTask),
+	}
+}
+
+// Offer offers a task to a potential consumer (poller)
+// If the task is successfully matched with a consumer, this
+// method will return true and no error. If the task is matched
+// but consumer returned error, then this method will return
+// true and error message. Both regular tasks and query tasks
+// should use this method to match with a consumer. Likewise, sync matches
+// and non-sync matches both should use this method.
+// returns error when:
+//  - ratelimit is exceeded (does not apply to query task)
+//  - context deadline is exceeded
+//  - task is matched and consumer returns error in response channel
+func (tm *TaskMatcher) Offer(ctx context.Context, task *internalTask) (bool, error) {
+	if task.isQuery() {
+		select {
+		case tm.queryTaskC <- task:
+			<-task.syncResponseCh
+			return true, nil
+		case <-ctx.Done():
+			return false, ctx.Err()
+		}
+	}
+
+	rsv, err := tm.ratelimit(ctx)
+	if err != nil {
+		tm.scope().IncCounter(metrics.SyncThrottleCounter)
+		return false, err
+	}
+
+	select {
+	case tm.taskC <- task: // poller picked up the task
+		if task.syncResponseCh != nil {
+			// if there is a response channel, block until resp is received
+			// and return error if the response contains error
+			err = <-task.syncResponseCh
+			return true, err
+		}
+		return false, err
+	default: // no poller waiting for tasks
+		if rsv != nil {
+			// there was a ratelimit token we consumed
+			// return it since we did not really do any work
+			rsv.Cancel()
+		}
+		return false, nil
+	}
+}
+
+// MustOffer blocks until a consumer is found to handle this task
+// Returns error only when context is canceled or the ratelimit is set to zero (allow nothing)
```

Comment:
Should it be rate exceeds threshold instead of ratelimit being zero?

Reply:
see comment below. this blocks until ratelimit token becomes available.

Reply:
I see. I misread the "context MUST NOT have a deadline associated"
In addition of this comment, maybe also add a check in the method in case someone misuse this method in some other place